File: year2025/day3/part1/solution.py
#!/usr/bin/env python3
# Day 3: Lobby, Part 1

FILE_PATH = "../.input"

def main():
    total_joltage = 0

    with open(FILE_PATH, 'r') as banks_file:
        for bank in banks_file:
            bank = bank.strip()
            # print(f"DEBUG: bank: {bank}")
            first_battery = int(bank[-2])
            second_battery = int(bank[-1])
            
            # This is a sort of ugly solution, but the logic is there
            # These issues may be addressed in part2, there's also a lazy solution that just imports combinations lol
            for battery in map(int, bank[:-2][::-1]):  # Reverse for bubble sort logic
                # print(f"DEBUG: battery: {battery}")
                if battery >= first_battery:
                    # print(f"DEBUG: replacing first_battery {first_battery} -> {battery}")
                    battery, first_battery = first_battery, battery
                else: continue
                if battery >= second_battery:
                    # print(f"DEBUG: replacing second_battery {second_battery} -> {battery}")
                    second_battery = battery
                # print("")
            
            combined_joltage = first_battery * 10 + second_battery
            # print(f"DEBUG: {bank}: {combined_joltage}")
            total_joltage += combined_joltage
    print(f"Total Joltage: {total_joltage}")

if __name__ == "__main__":
    main()

File: year2025/day3/part1/solution.cpp
// Day 3: Lobby, Part 1

#include <iostream>
#include <fstream>
#include <string>
#include <algorithm>

using namespace std;

const string FILE_PATH = "../.input";

int main() {
    unsigned long long total_joltage = 0;
    ifstream banks_file(FILE_PATH);
    string bank;

    while (getline(banks_file, bank)) {
        //string orig_bank = bank;  // Debug group
        // These are chars for display purposes, but behave the same with the exception of the ascii conversion when computing combined_joltage.
        char second_battery = bank.back();
        bank.pop_back();
        char first_battery = bank.back();
        bank.pop_back();

        reverse(bank.begin(), bank.end());  // Reverse shortened list for bubble sort type logic

        for (char battery : bank) {
            //cout << "DEBUG: battery: " << battery << endl;
            if (battery >= first_battery) {
                //cout << "DEBUG: replacing first_battery " << first_battery << " -> " << battery << endl;
                char temp_battery = first_battery;
                first_battery = battery;
                battery = temp_battery;
            } else { continue; }
            if (battery >= second_battery) {
                //cout << "DEBUG: replacing second_battery " << second_battery << " -> " << battery << endl;
                second_battery = battery;
            }
        }

        unsigned combined_joltage = (first_battery - '0')*10 + (second_battery - '0');
        //cout << "DEBUG: " << (first_battery - '0')*10 << ", " << (second_battery - '0') << endl;
        //cout << "DEBUG: " << orig_bank << ": " << combined_joltage << endl;  // Debug group
        total_joltage += combined_joltage;
    }

    cout << "Total Joltage: " << total_joltage << endl;
    return 0;
}
